src/bootstrap/builder.rs

```diff
@@ -1090,35 +1090,32 @@ impl<'a> Builder<'a> {
             rustflags.arg("-Zunstable-options");
         }
 
-        // #[cfg(not(bootstrap)]
-        if stage != 0 {
-            // Enable cfg checking of cargo features
-            // FIXME: De-comment this when cargo beta get support for it
-            // cargo.arg("-Zcheck-cfg-features");
-
-            // Enable cfg checking of rustc well-known names
-            rustflags
-                .arg("-Zunstable-options")
-                // Enable checking of well known names
-                .arg("--check-cfg=names()")
-                // Enable checking of well known values
-                .arg("--check-cfg=values()");
-
-            // Add extra cfg not defined in rustc
-            for (restricted_mode, name, values) in EXTRA_CHECK_CFGS {
-                if *restricted_mode == None || *restricted_mode == Some(mode) {
-                    // Creating a string of the values by concatenating each value:
-                    // ',"tvos","watchos"' or '' (nothing) when there are no values
-                    let values = match values {
-                        Some(values) => values
-                            .iter()
-                            .map(|val| [",", "\"", val, "\""])
-                            .flatten()
-                            .collect::<String>(),
-                        None => String::new(),
-                    };
-                    rustflags.arg(&format!("--check-cfg=values({name}{values})"));
-                }
+        // Enable cfg checking of cargo features
+        // FIXME: De-comment this when cargo beta get support for it
+        // cargo.arg("-Zcheck-cfg-features");
+
+        // Enable cfg checking of rustc well-known names
+        rustflags
+            .arg("-Zunstable-options")
+            // Enable checking of well known names
+            .arg("--check-cfg=names()")
+            // Enable checking of well known values
+            .arg("--check-cfg=values()");
```

Comment:
This is causing a failure when compiling `rustc_llvm`, and I'm not sure how to best handle it. The build script of that crate sets the `llvm_component` for every LLVM component enabled as part of the build:

https://github.com/rust-lang/rust/blob/634770c0a7f8598164ab825cfe419cc8b03c36e5/compiler/rustc_llvm/build.rs#L175-L177

The list of available components is also quite long:

https://github.com/rust-lang/rust/blob/634770c0a7f8598164ab825cfe419cc8b03c36e5/compiler/rustc_llvm/build.rs#L134-L163

My understanding is that, since some of those flags won't be passed, we need to add them to the list of allowed flags:

https://github.com/rust-lang/rust/blob/634770c0a7f8598164ab825cfe419cc8b03c36e5/src/bootstrap/lib.rs#L191-L225

I'm worried that the list is quite long though, and it might be annoying keeping it up to date. Is there a better way?

cc @Urgau

Reply:
Oh, sorry for that.

> I'm worried that the list is quite long though, and it might be annoying keeping it up to date. Is there a better way?

I think it could be done directly in the `build.rs` instead of adding it to the global extra. Something like this might work:

```rust
println!("cargo:rustc-flags=-Zunstable-options");
println!("cargo:rustc-flags=--check-cfg=values(\"llvm_component\", ...)");
```

Reply:
Okay, it doesn't seems possible to set custom `rustflags` in build script so I would advise to temporarily not enable well known values checking in stage 0:
```rust
// FIXME(Urgau): Enabling this in stage 0 break `rustc_llvm`, should be fixed when
// cargo has build script support for --check-cfg
if stage != 0 {
   // Enable checking of well known values
   rustflags.arg("--check-cfg=values()");
}
```
Or not enable at all cfg checking in stage 0 like it's currently done, but this seems excessive and unnecessary.

Reply:
Ok applied your suggestion here, thanks!

Reply:
@pietroalbini The new build errors are because some extra `target_env` and `target_os` and defined in `EXTRA_CHECK_CFGS` and are added, but because we now disable checking of well known values in stage 0 only those extra values are defined leading to the new build failure on valid config.

Adapting the existing code to take this into account doesn't seems this PR job, so I recommend to not enable any cfg checking at stage 0, ie let everything like it was. I will do a follow up PR fixing this and adding the cargo `-Zcheck-cfg-features`.

Reply:
Ok reverted the whole enabing of `--check-cfg` for stage0.